File: android/app/src/test/java/com/buttonlog/app/network/NetworkErrorTests.kt
package com.buttonlog.app.network

import com.buttonlog.app.data.model.ApiError
import com.google.common.truth.Truth.assertThat
import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.ResponseBody.Companion.toResponseBody
import org.junit.Test
import retrofit2.HttpException
import retrofit2.Response
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException

/**
 * Tests for network error handling scenarios.
 *
 * These tests verify the app handles various network failure conditions:
 * - Connection timeouts
 * - Server errors (5xx)
 * - Malformed responses
 * - Network connectivity issues
 */
class NetworkErrorTests {

    private val gson = Gson()

    // Test data class for API response parsing
    data class TestApiResponse(
        val success: Boolean = false,
        val data: Any? = null,
        val error: ApiError? = null
    )

    // MARK: - Timeout Tests

    @Test
    fun `timeout exception is identifiable`() {
        val exception = SocketTimeoutException("Connection timed out")

        assertThat(exception).isInstanceOf(IOException::class.java)
        assertThat(isRetryableException(exception)).isTrue()
    }

    @Test
    fun `unknown host exception is identifiable`() {
        val exception = UnknownHostException("Unable to resolve host")

        assertThat(exception).isInstanceOf(IOException::class.java)
        assertThat(isRetryableException(exception)).isFalse()
    }

    // MARK: - Malformed Response Tests

    @Test
    fun `malformed JSON throws JsonSyntaxException`() {
        val malformedJson = "{ invalid json }"

        try {
            gson.fromJson(malformedJson, TestApiResponse::class.java)
            throw AssertionError("Should have thrown JsonSyntaxException")
        } catch (e: JsonSyntaxException) {
            // Expected
            assertThat(e).isInstanceOf(JsonSyntaxException::class.java)
        }
    }

    @Test
    fun `empty response parses as null`() {
        val emptyJson = ""

        val response = gson.fromJson(emptyJson, TestApiResponse::class.java)

        // Gson returns null for empty string
        assertThat(response).isNull()
    }

    @Test
    fun `unexpected structure parses with null fields`() {
        val unexpectedJson = """
            {
                "unexpected_field": "value",
                "another_field": 123
            }
        """.trimIndent()

        val response = gson.fromJson(unexpectedJson, TestApiResponse::class.java)

        // Fields should be null/default when not present
        assertThat(response.success).isFalse()
        assertThat(response.data).isNull()
    }

    // MARK: - HTTP Error Code Tests

    @Test
    fun `unauthorized error parses correctly`() {
        val errorJson = """
            {
                "success": false,
                "error": {
                    "code": "UNAUTHORIZED",
                    "message": "Authentication required"
                }
            }
        """.trimIndent()

        val response = gson.fromJson(errorJson, TestApiResponse::class.java)

        assertThat(response.success).isFalse()
        assertThat(response.error).isNotNull()
        assertThat(response.error?.code).isEqualTo("UNAUTHORIZED")
        assertThat(response.error?.message).isEqualTo("Authentication required")
    }

    @Test
    fun `validation error parses correctly`() {
        val errorJson = """
            {
                "success": false,
                "error": {
                    "code": "VALIDATION_ERROR",
                    "message": "Email is invalid"
                }
            }
        """.trimIndent()

        val response = gson.fromJson(errorJson, TestApiResponse::class.java)

        assertThat(response.success).isFalse()
        assertThat(response.error?.code).isEqualTo("VALIDATION_ERROR")
        assertThat(response.error?.message).contains("invalid")
    }

    @Test
    fun `server error parses correctly`() {
        val errorJson = """
            {
                "success": false,
                "error": {
                    "code": "INTERNAL_SERVER_ERROR",
                    "message": "Something went wrong"
                }
            }
        """.trimIndent()

        val response = gson.fromJson(errorJson, TestApiResponse::class.java)

        assertThat(response.success).isFalse()
        assertThat(response.error?.code).isEqualTo("INTERNAL_SERVER_ERROR")
    }

    @Test
    fun `rate limited error parses correctly`() {
        val errorJson = """
            {
                "success": false,
                "error": {
                    "code": "RATE_LIMITED",
                    "message": "Too many requests"
                }
            }
        """.trimIndent()

        val response = gson.fromJson(errorJson, TestApiResponse::class.java)

        assertThat(response.success).isFalse()
        assertThat(response.error?.code).isEqualTo("RATE_LIMITED")
    }

    // MARK: - HTTP Exception Tests

    @Test
    fun `HttpException with 401 is unauthorized`() {
        val response = Response.error<Any>(
            401,
            "Unauthorized".toResponseBody("application/json".toMediaType())
        )
        val exception = HttpException(response)

        assertThat(exception.code()).isEqualTo(401)
        assertThat(isUnauthorizedException(exception)).isTrue()
    }

    @Test
    fun `HttpException with 404 is not found`() {
        val response = Response.error<Any>(
            404,
            "Not Found".toResponseBody("application/json".toMediaType())
        )
        val exception = HttpException(response)

        assertThat(exception.code()).isEqualTo(404)
        assertThat(isNotFoundError(exception)).isTrue()
    }

    @Test
    fun `HttpException with 500 is server error`() {
        val response = Response.error<Any>(
            500,
            "Internal Server Error".toResponseBody("application/json".toMediaType())
        )
        val exception = HttpException(response)

        assertThat(exception.code()).isEqualTo(500)
        assertThat(isServerError(exception)).isTrue()
    }

    @Test
    fun `HttpException with 503 is retryable`() {
        val response = Response.error<Any>(
            503,
            "Service Unavailable".toResponseBody("application/json".toMediaType())
        )
        val exception = HttpException(response)

        assertThat(exception.code()).isEqualTo(503)
        assertThat(isRetryableHttpException(exception)).isTrue()
    }

    // MARK: - Retry Logic Tests

    @Test
    fun `retryable exceptions are identified correctly`() {
        val retryableExceptions = listOf(
            SocketTimeoutException("timeout"),
            IOException("Network unreachable")
        )

        val nonRetryableExceptions = listOf(
            UnknownHostException("unknown host"),
            IllegalArgumentException("bad argument"),
            NullPointerException("null")
        )

        retryableExceptions.forEach { exception ->
            assertThat(isRetryableException(exception)).isTrue()
        }

        nonRetryableExceptions.forEach { exception ->
            assertThat(isRetryableException(exception)).isFalse()
        }
    }

    // MARK: - HTTP Status Code Tests

    @Test
    fun `success status codes are identified correctly`() {
        val successCodes = listOf(200, 201, 204)
        val errorCodes = listOf(400, 401, 403, 404, 422, 429, 500, 502, 503, 504)

        successCodes.forEach { code ->
            assertThat(isSuccessStatusCode(code)).isTrue()
        }

        errorCodes.forEach { code ->
            assertThat(isSuccessStatusCode(code)).isFalse()
        }
    }

    @Test
    fun `client error status codes are identified correctly`() {
        val clientErrorCodes = listOf(400, 401, 403, 404, 422, 429)
        val nonClientErrorCodes = listOf(200, 201, 500, 503)

        clientErrorCodes.forEach { code ->
            assertThat(isClientError(code)).isTrue()
        }

        nonClientErrorCodes.forEach { code ->
            assertThat(isClientError(code)).isFalse()
        }
    }

    @Test
    fun `server error status codes are identified correctly`() {
        val serverErrorCodes = listOf(500, 502, 503, 504)
        val nonServerErrorCodes = listOf(200, 201, 400, 401, 404)

        serverErrorCodes.forEach { code ->
            assertThat(isServerErrorCode(code)).isTrue()
        }

        nonServerErrorCodes.forEach { code ->
            assertThat(isServerErrorCode(code)).isFalse()
        }
    }

    // MARK: - Error Message Extraction

    @Test
    fun `error message extracts from API response`() {
        val errorJson = """
            {
                "success": false,
                "error": {
                    "code": "VALIDATION_ERROR",
                    "message": "Email is required"
                }
            }
        """.trimIndent()

        val message = extractErrorMessage(errorJson)

        assertThat(message).isEqualTo("Email is required")
    }

    @Test
    fun `error message returns default for invalid JSON`() {
        val invalidJson = "not json"

        val message = extractErrorMessage(invalidJson)

        assertThat(message).isEqualTo("An unknown error occurred")
    }

    // MARK: - Helper Methods

    private fun isRetryableException(exception: Exception): Boolean {
        return when (exception) {
            is SocketTimeoutException -> true
            is IOException -> exception !is UnknownHostException
            else -> false
        }
    }

    private fun isRetryableHttpException(exception: HttpException): Boolean {
        return exception.code() in listOf(408, 429, 500, 502, 503, 504)
    }

    private fun isUnauthorizedException(exception: HttpException): Boolean {
        return exception.code() == 401
    }

    private fun isNotFoundError(exception: HttpException): Boolean {
        return exception.code() == 404
    }

    private fun isServerError(exception: HttpException): Boolean {
        return exception.code() >= 500
    }

    private fun isSuccessStatusCode(code: Int): Boolean {
        return code in 200..299
    }

    private fun isClientError(code: Int): Boolean {
        return code in 400..499
    }

    private fun isServerErrorCode(code: Int): Boolean {
        return code in 500..599
    }

    private fun extractErrorMessage(responseBody: String): String {
        return try {
            val response = gson.fromJson(responseBody, TestApiResponse::class.java)
            response.error?.message ?: "An unknown error occurred"
        } catch (e: Exception) {
            "An unknown error occurred"
        }
    }
}

File: iphone/ButtonLogTests/Integration/NetworkErrorTests.swift
import XCTest
@testable import ButtonLog

/// Tests for network error handling scenarios.
///
/// These tests verify the app handles various network failure conditions:
/// - Connection timeouts
/// - Server errors (5xx)
/// - Malformed responses
/// - Network connectivity issues
final class NetworkErrorTests: XCTestCase {

    var session: URLSession!

    override func setUp() async throws {
        try await super.setUp()

        // Create session with short timeout for testing
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 5.0
        config.timeoutIntervalForResource = 10.0
        session = URLSession(configuration: config)
    }

    override func tearDown() async throws {
        session = nil
        try await super.tearDown()
    }

    // MARK: - Timeout Tests

    func testRequest_withTimeout_handlesGracefully() async throws {
        // Use a non-routable IP to force timeout
        let url = URL(string: "http://10.255.255.1/api/test")!
        var request = URLRequest(url: url)
        request.timeoutInterval = 2.0 // Short timeout

        do {
            let _ = try await session.data(for: request)
            XCTFail("Should have timed out")
        } catch let error as URLError {
            // Expected timeout error
            XCTAssertTrue(
                error.code == .timedOut || error.code == .cannotConnectToHost,
                "Expected timeout or connection error, got: \(error.code)"
            )
            print("Got expected timeout error: \(error.localizedDescription)")
        } catch {
            print("Got error: \(error)")
            // Other network errors are acceptable
        }
    }

    func testRequest_withInvalidHost_handlesGracefully() async throws {
        let url = URL(string: "http://invalid.host.that.does.not.exist.buttonlog.test/api/test")!

        do {
            let _ = try await session.data(for: URLRequest(url: url))
            XCTFail("Should have failed with invalid host")
        } catch let error as URLError {
            XCTAssertTrue(
                error.code == .cannotFindHost || error.code == .cannotConnectToHost,
                "Expected host resolution error, got: \(error.code)"
            )
            print("Got expected DNS error: \(error.localizedDescription)")
        } catch {
            print("Got error: \(error)")
        }
    }

    // MARK: - Malformed Response Tests

    func testAPIService_withMalformedJSON_handlesGracefully() async throws {
        // Test that the decoder handles malformed JSON properly
        let malformedJSON = "{ invalid json }"
        let data = malformedJSON.data(using: .utf8)!

        do {
            let _ = try JSONDecoder().decode(IntegrationAPIResponse<[String]>.self, from: data)
            XCTFail("Should have thrown decoding error")
        } catch is DecodingError {
            print("Got expected decoding error for malformed JSON")
        }
    }

    func testAPIService_withEmptyResponse_handlesGracefully() async throws {
        let emptyData = Data()

        do {
            let _ = try JSONDecoder().decode(IntegrationAPIResponse<[String]>.self, from: emptyData)
            XCTFail("Should have thrown decoding error")
        } catch is DecodingError {
            print("Got expected decoding error for empty response")
        }
    }

    func testAPIService_withUnexpectedStructure_handlesGracefully() async throws {
        // Response with different structure than expected
        let unexpectedJSON = """
        {
            "unexpected_field": "value",
            "another_field": 123
        }
        """.data(using: .utf8)!

        do {
            let response = try JSONDecoder().decode(IntegrationAPIResponse<[String]>.self, from: unexpectedJSON)
            // If it decodes, success should be false and data should be nil
            XCTAssertFalse(response.success)
        } catch is DecodingError {
            print("Got expected decoding error for unexpected structure")
        }
    }

    // MARK: - HTTP Error Code Tests

    func testAPIError_unauthorized_parsesCorrectly() async throws {
        let errorJSON = """
        {
            "success": false,
            "error": {
                "code": "UNAUTHORIZED",
                "message": "Authentication required"
            }
        }
        """.data(using: .utf8)!

        let response = try JSONDecoder().decode(IntegrationErrorResponse.self, from: errorJSON)

        XCTAssertFalse(response.success)
        XCTAssertEqual(response.error?.code, "UNAUTHORIZED")
        XCTAssertEqual(response.error?.message, "Authentication required")
    }

    func testAPIError_validationError_parsesCorrectly() async throws {
        let errorJSON = """
        {
            "success": false,
            "error": {
                "code": "VALIDATION_ERROR",
                "message": "Email is invalid",
                "details": {
                    "email": ["is invalid format"]
                }
            }
        }
        """.data(using: .utf8)!

        let response = try JSONDecoder().decode(IntegrationErrorResponse.self, from: errorJSON)

        XCTAssertFalse(response.success)
        XCTAssertEqual(response.error?.code, "VALIDATION_ERROR")
        XCTAssertTrue(response.error?.message?.contains("invalid") ?? false)
    }

    func testAPIError_serverError_parsesCorrectly() async throws {
        let errorJSON = """
        {
            "success": false,
            "error": {
                "code": "INTERNAL_SERVER_ERROR",
                "message": "Something went wrong"
            }
        }
        """.data(using: .utf8)!

        let response = try JSONDecoder().decode(IntegrationErrorResponse.self, from: errorJSON)

        XCTAssertFalse(response.success)
        XCTAssertEqual(response.error?.code, "INTERNAL_SERVER_ERROR")
    }

    func testAPIError_rateLimited_parsesCorrectly() async throws {
        let errorJSON = """
        {
            "success": false,
            "error": {
                "code": "RATE_LIMITED",
                "message": "Too many requests. Please try again later."
            }
        }
        """.data(using: .utf8)!

        let response = try JSONDecoder().decode(IntegrationErrorResponse.self, from: errorJSON)

        XCTAssertFalse(response.success)
        XCTAssertEqual(response.error?.code, "RATE_LIMITED")
    }

    // MARK: - Retry Logic Tests

    func testRetryableError_identifiesCorrectly() {
        // Test which errors should trigger retry
        let retryableErrors: [URLError.Code] = [
            .timedOut,
            .networkConnectionLost,
            .notConnectedToInternet
        ]

        let nonRetryableErrors: [URLError.Code] = [
            .badURL,
            .unsupportedURL,
            .cannotFindHost,
            .cancelled
        ]

        for code in retryableErrors {
            let error = URLError(code)
            XCTAssertTrue(isRetryableError(error), "\(code) should be retryable")
        }

        for code in nonRetryableErrors {
            let error = URLError(code)
            XCTAssertFalse(isRetryableError(error), "\(code) should not be retryable")
        }
    }

    // MARK: - Connection State Tests

    func testConnectionState_offline_handlesGracefully() async throws {
        // Simulate offline by using invalid URL
        let url = URL(string: "http://0.0.0.0:1/api/test")!

        do {
            let _ = try await session.data(for: URLRequest(url: url))
            XCTFail("Should have failed when offline")
        } catch {
            print("Got expected error for offline state: \(error.localizedDescription)")
            // Any network error is acceptable here
        }
    }

    // MARK: - Helper Methods

    private func isRetryableError(_ error: URLError) -> Bool {
        switch error.code {
        case .timedOut,
             .networkConnectionLost,
             .notConnectedToInternet:
            return true
        default:
            return false
        }
    }
}

// MARK: - Mock Server Response Tests

extension NetworkErrorTests {

    /// Tests that the app correctly handles various HTTP status codes
    func testHTTPStatusCode_handling() {
        // 2xx - Success
        XCTAssertTrue(isSuccessStatusCode(200))
        XCTAssertTrue(isSuccessStatusCode(201))
        XCTAssertTrue(isSuccessStatusCode(204))

        // 4xx - Client errors
        XCTAssertFalse(isSuccessStatusCode(400)) // Bad Request
        XCTAssertFalse(isSuccessStatusCode(401)) // Unauthorized
        XCTAssertFalse(isSuccessStatusCode(403)) // Forbidden
        XCTAssertFalse(isSuccessStatusCode(404)) // Not Found
        XCTAssertFalse(isSuccessStatusCode(422)) // Unprocessable Entity
        XCTAssertFalse(isSuccessStatusCode(429)) // Too Many Requests

        // 5xx - Server errors
        XCTAssertFalse(isSuccessStatusCode(500)) // Internal Server Error
        XCTAssertFalse(isSuccessStatusCode(502)) // Bad Gateway
        XCTAssertFalse(isSuccessStatusCode(503)) // Service Unavailable
        XCTAssertFalse(isSuccessStatusCode(504)) // Gateway Timeout
    }

    private func isSuccessStatusCode(_ code: Int) -> Bool {
        return code >= 200 && code < 300
    }
}
